### ISSUE TITLE
Adding WebGL texture types

### ISSUE THREAD

**Issue by RandomGamingDev** [processing/p5.js#6166]
### Increasing Access

p5.js's addition of shaders meant that a lot more was possible with the library, but the lack of precise texture control means that people either have to waste memory, speed, or themselves poke into the WebGL and p5.js code to get all the data needed for proper creation of textures. For example, if someone wanted 1d or 3d textures, or if they wanted to use UNSIGNED_BYTE textures.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [X] WebGL
- [ ] Build Process
- [ ] Unit Testing
- [ ] Internalization
- [ ] Friendly Errors
- [ ] Other (specify if possible)

### Feature request details

Give a function similar to the `createImage` function for WebGL only that gives control over the WebGL parameters of the texture while also working with the other parts of the p5 library.

**Comment by RandomGamingDev** [comment]
Nice! For now I think I might just make my own p5.js addon since it sounds like it's only gonna come during the summer.

**Comment by davepagurek** [comment]
awesome, feel free to ask on the p5 discord if you need any pointers on making that happen!

**Comment by RandomGamingDev** [comment]
Here's the library for anyone coming across this: https://github.com/RandomGamingDev/WebGL2Tex

**Comment by Vishal2002** [comment]
Can I also work on this issue ?

**Comment by RandomGamingDev** [comment]
@Vishal2002 Sure, but you'll probably want to ask for someone to assign you to this issue. @davepagurek could you help with that?

**Comment by davepagurek** [comment]
Hi, thanks for your interest @Vishal2002! Before getting started, maybe we should iron out what specifically this request involves. @RandomGamingDev mentioned [3D textures](https://developer.mozilla.org/en-US/docs/Web/API/WebGL2RenderingContext/texImage3D) and more texture formats. Either probably requires a bit of discussion of how to add them in, and maybe a motivating use of each that we would use as an example and would be used to guide the discussion of implementation.

**Comment by Vishal2002** [comment]
> @Vishal2002 Sure, but you'll probably want to ask for someone to assign you to this issue. @davepagurek could you help with that?

I want to work on this issue and learn more about the requirements, I know the basics of WEBGL like Camera, Texture and all.Should we discuss that on Discord or here.

**Comment by davepagurek** [comment]
> I want to work on this issue and learn more about the requirements, I know the basics of WEBGL like Camera, Texture and all.Should we discuss that on Discord or here.

For each feature being added, probably we'll want a use in mind that motivates its inclusion, to guide the discussion and use as examples. Maybe Discord could be a good spot to brainstorm those before bringing them to Github to discuss?

**Comment by RandomGamingDev** [comment]
> For each feature being added, probably we'll want a use in mind that motivates its inclusion, to guide the discussion and use as examples. Maybe Discord could be a good spot to brainstorm those before bringing them to Github to discuss?

That sounds great. I'd also be happy to talk about it with others on Discord :D

**Comment by ksen0** [comment]
HI @RandomGamingDev @davepagurek ! While reviewing the [open issues in the WebGL project board](https://github.com/orgs/processing/projects/20/views/1) I saw this one. Now that p5.js 2.0 is live and feature-frozen, I think the options for this idea are:

1. Open a wider discussion of what is proposed here, as Dave suggested, for consideration in 2.x. Currently there's more of a focus on the new [p5.strands](https://beta.p5js.org/tutorials/intro-to-p5-strands/) feature that aims to make shaders more beginner-friendly, but shaders are still supported as well, and it's also important to support the transition from p5.strands to GLSL for those who are interested and want more control.
2. Focus this issue on [a new add-on library](https://github.com/processing/p5.js-addon-template) that can be more experimental and exploratory - but inviting contribution from the community

cc @lukeplowden in case you have thoughts on this too!
What do you think?

**Comment by lukeplowden** [comment]
There is quite a lot to decide in terms of how much to expose and how to expose it. I think it could be great to have 3D textures in p5.strands. There could be a variant of filter shaders to operate on them and draw volumes. I've been thinking about how TouchDesigner makes operations on textures so easy, so I was trying to layer some filters in a similar way. Warning, this sketch can flash: [https://editor.p5js.org/lukeplowden/sketches/YJHx8kt9g ](https://editor.p5js.org/lukeplowden/sketches/YJHx8kt9g )

Right now, layering frame buffers & 3 calls to filter in this sketch (although I did similar in the p5.strands tutorial) is not super intuitive in the API. If we add new texture options, we should also consider ways of operating on them. For example, could we have a `volumeShader` which lets us layer different volume 'filters' in a series, and make the `filterShader` hooks feel similar? 

In the sketch above, I layered 3 filter calls and essentially use 3 framebuffers, because I was lazy to write a blur shader when p5.js already has one. But I think this is a little bit advanced usage. It could be that you could use filters within strands:
```js
    let col = getTexture(canvasContent, inputs.texCoord);
    let blur1 = blur(col, 2);
    let blur2 = blur(blur1, 10);
```
Then, for 3D or 1D textures, there could be different abstractions, but maybe a `volumeShader` could be a start for the 3D textures. I think expanded options for textures would be great, these are just my initial thoughts. I'm not thinking about making all GL texture options open to the user, but about how they could be used interestingly, and how to make everything in strands generally cohesive with the rest of the library.

**Comment by RandomGamingDev** [comment]
> HI [@RandomGamingDev](https://github.com/RandomGamingDev) [@davepagurek](https://github.com/davepagurek) ! While reviewing the [open issues in the WebGL project board](https://github.com/orgs/processing/projects/20/views/1) I saw this one. Now that p5.js 2.0 is live and feature-frozen, I think the options for this idea are:
> 
>     1. Open a wider discussion of what is proposed here, as Dave suggested, for consideration in 2.x. Currently there's more of a focus on the new [p5.strands](https://beta.p5js.org/tutorials/intro-to-p5-strands/) feature that aims to make shaders more beginner-friendly, but shaders are still supported as well, and it's also important to support the transition from p5.strands to GLSL for those who are interested and want more control.
> 
>     2. Focus this issue on [a new add-on library](https://github.com/processing/p5.js-addon-template) that can be more experimental and exploratory - but inviting contribution from the community
> 
> 
> cc [@lukeplowden](https://github.com/lukeplowden) in case you have thoughts on this too! What do you think?

I agree with @lukeplowden here. I think it'd be great to open a wider discussion, especially since the potential when combined with p5.strands would be great.
